sql/src/main/java/io/crate/execution/engine/window/WindowBatchIterator.java

```diff
@@ -182,16 +213,28 @@ private void computeAndFillStandaloneOutgoingCellsFor(Row sourceRow) {
         }
     }
 
-    private void collectPeers() {
-        Iterable<Row> rows = peerRows.stream().collect(collector);
-        peerRows.clear();
-        peersCollected = true;
-        Row collectedRow = rows.iterator().next();
-        collectorNumberOfColumns = collectedRow.numColumns();
-        for (int i = 0; i < collectorNumberOfColumns; i++) {
-            outgoingCells[i] = collectedRow.get(i);
+    private void executeWindowFunctions() {
+        int startPosition = windowRowPosition;
+        WindowFrameState currentFrame = new WindowFrameState(
+            startPosition,
+            startPosition + windowForCurrentRow.size(),
+            windowForCurrentRow
+        );
+
+        Object[][] cellsForCurrentFrame = new Object[windowForCurrentRow.size()][windowFunctionsCount];
+
+        for (int i = 0; i < windowForCurrentRow.size(); i++) {
+            for (int funcIdx = 0; funcIdx < functions.size(); funcIdx++) {
+                WindowFunction function = functions.get(funcIdx);
+                Row result = function.execute(windowRowPosition + i, currentFrame);
```

Comment:
Could we just have a `execute :: Int -> WindowFrameState -> Object` ? We're always returning `Row1` and doing a `.materialize()[0]` here without checking the number of columns.

Reply:
yes, that's a great idea